dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java

```diff
@@ -519,6 +532,56 @@ else if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.CHILDREN
         return sql;
     }
 
+    /**
+     * This method will generate a sql sentence responsible for filtering only the
+     * allowed category options for this program, and expects to receive only the
+     * authorized category options.
+     *
+     * @see org.hisp.dhis.analytics.AnalyticsSecurityManager#decideAccess(DataQueryParams)
+     *
+     * @param programCategories the list of program categories containing the list
+     *        of authorized category options for the current user.
+     * @return the sql statement in the format:
+     *          "and ax."mEXqxV2KIUl" in ('qNqYLugIySD') or ax."r7NDRdgj5zs" in ('qNqYLugIySD') "
+     */
+    String queryOnlyAllowedCategoryOptionEvents( final List<Category> programCategories )
+    {
+        boolean andFlag = true;
+        final StringBuilder query = new StringBuilder();
```

Comment:
We have a util class `SqlHelper`, you can add a "andOr" method there to encapsulate the and/or logic.

Reply:
Changes applied. Waiting Trevis to build.